Build error on Linux

I am seeing the following failure when building on Linux starting from 6d68373300d230166040c3ae261e26ea17701652. The following is the critical part of the compilation error I am seeing.

```
/home/quinn/nodwork/onnx-mlir/src/Builder/FrontendDialectHelper.cpp:155:27: error: no matching constructor for initialization of 'ArrayRefOrSmallVector' (aka 'llvm::ArrayRef<float>')
    ArrayRefOrSmallVector array(data.begin(), data.end());
                          ^     ~~~~~~~~~~~~~~~~~~~~~~~~
/home/quinn/nodwork/onnx-mlir/src/Builder/FrontendDialectHelper.cpp:187:12: note: in instantiation of function template specialization 'onnx_mlir::createDenseElmAttr<float>' requested here
    return createDenseElmAttr<float>(externalDataDir, tp, denseBuilder);
           ^
/home/quinn/nodwork/onnx-mlir/llvm-project/llvm/include/llvm/ADT/ArrayRef.h:78:18: note: candidate constructor not viable: no known conversion from 'google::protobuf::RepeatedField<float>::const_iterator' (aka 'RepeatedIterator<const float>') to 'const float *' for 1st argument
    /*implicit*/ ArrayRef(const T *data, size_t length)
                 ^
/home/quinn/nodwork/onnx-mlir/llvm-project/llvm/include/llvm/ADT/ArrayRef.h:82:5: note: candidate constructor not viable: no known conversion from 'google::protobuf::RepeatedField<float>::const_iterator' (aka 'RepeatedIterator<const float>') to 'const float *' for 1st argument
    ArrayRef(const T *begin, const T *end)
    ^
/home/quinn/nodwork/onnx-mlir/llvm-project/llvm/include/llvm/ADT/ArrayRef.h:125:5: note: candidate template ignored: could not match 'ArrayRef' against 'RepeatedIterator'
    ArrayRef(const ArrayRef<U *> &A,
    ^
/home/quinn/nodwork/onnx-mlir/llvm-project/llvm/include/llvm/ADT/ArrayRef.h:134:18: note: candidate template ignored: could not match 'SmallVectorTemplateCommon' against 'RepeatedIterator'
    /*implicit*/ ArrayRef(
                 ^
/home/quinn/nodwork/onnx-mlir/llvm-project/llvm/include/llvm/ADT/ArrayRef.h:143:5: note: candidate template ignored: could not match 'vector' against 'RepeatedIterator'
    ArrayRef(const std::vector<U *, A> &Vec,
    ^
/home/quinn/nodwork/onnx-mlir/llvm-project/llvm/include/llvm/ADT/ArrayRef.h:71:18: note: candidate constructor not viable: requires 1 argument, but 2 were provided
    /*implicit*/ ArrayRef(NoneType) {}
                 ^
/home/quinn/nodwork/onnx-mlir/llvm-project/llvm/include/llvm/ADT/ArrayRef.h:74:18: note: candidate constructor not viable: requires single argument 'OneElt', but 2 arguments were provided
    /*implicit*/ ArrayRef(const T &OneElt)
                 ^
/home/quinn/nodwork/onnx-mlir/llvm-project/llvm/include/llvm/ADT/ArrayRef.h:89:18: note: candidate constructor template not viable: requires single argument 'Vec', but 2 arguments were provided
    /*implicit*/ ArrayRef(const SmallVectorTemplateCommon<T, U> &Vec)
                 ^
/home/quinn/nodwork/onnx-mlir/llvm-project/llvm/include/llvm/ADT/ArrayRef.h:95:18: note: candidate constructor template not viable: requires single argument 'Vec', but 2 arguments were provided
    /*implicit*/ ArrayRef(const std::vector<T, A> &Vec)
                 ^
/home/quinn/nodwork/onnx-mlir/llvm-project/llvm/include/llvm/ADT/ArrayRef.h:100:28: note: candidate constructor template not viable: requires single argument 'Arr', but 2 arguments were provided
    /*implicit*/ constexpr ArrayRef(const std::array<T, N> &Arr)
                           ^
/home/quinn/nodwork/onnx-mlir/llvm-project/llvm/include/llvm/ADT/ArrayRef.h:105:28: note: candidate constructor template not viable: requires single argument 'Arr', but 2 arguments were provided
    /*implicit*/ constexpr ArrayRef(const T (&Arr)[N]) : Data(Arr), Length(N) {}
                           ^
/home/quinn/nodwork/onnx-mlir/llvm-project/llvm/include/llvm/ADT/ArrayRef.h:115:18: note: candidate constructor not viable: requires single argument 'Vec', but 2 arguments were provided
    /*implicit*/ ArrayRef(const std::initializer_list<T> &Vec)
                 ^
/home/quinn/nodwork/onnx-mlir/llvm-project/llvm/include/llvm/ADT/ArrayRef.h:42:40: note: candidate constructor (the implicit copy constructor) not viable: requires 1 argument, but 2 were provided
  class LLVM_GSL_POINTER [[nodiscard]] ArrayRef {
                                       ^
/home/quinn/nodwork/onnx-mlir/llvm-project/llvm/include/llvm/ADT/ArrayRef.h:42:40: note: candidate constructor (the implicit move constructor) not viable: requires 1 argument, but 2 were provided
/home/quinn/nodwork/onnx-mlir/llvm-project/llvm/include/llvm/ADT/ArrayRef.h:68:18: note: candidate constructor not viable: requires 0 arguments, but 2 were provided
    /*implicit*/ ArrayRef() = default;

```

I am on LLVM commit hash 59548fe873d8d98e359fb21fbb2a0852fed17ff5. Please let me know if there is anything else needed from me to reproduce and/or fix the error.

--- a/src/Builder/FrontendDialectHelper.cpp
+++ b/src/Builder/FrontendDialectHelper.cpp
@@ -122,7 +122,7 @@ template <typename T, typename U>
 std::enable_if_t<std::is_same_v<T, U>, mlir::DenseElementsAttr>
 createDenseElmAttrFromProtoData(const google::protobuf::RepeatedField<U> &data,
     DenseElementsAttrBuilder<T> denseBuilder) {
-  return denseBuilder(llvm::makeArrayRef(data.begin(), data.end()));
+  return denseBuilder(llvm::makeArrayRef(data.data(), data.size()));
 }
 
 // When the protobuf repeated field has a type larger than T,

--- a/src/Builder/FrontendDialectHelper.cpp
+++ b/src/Builder/FrontendDialectHelper.cpp
@@ -116,6 +116,26 @@ template <typename T>
 using DenseElementsAttrBuilder =
     llvm::function_ref<mlir::DenseElementsAttr(llvm::ArrayRef<T>)>;
 
+// When the protobuf repeated field has a type of the same size as T,
+// access the data directly via ArrayRef.
+template <typename T, typename U>
+std::enable_if_t<std::is_same_v<T, U>, mlir::DenseElementsAttr>
+createDenseElmAttrFromProtoData(const google::protobuf::RepeatedField<U> &data,
+    DenseElementsAttrBuilder<T> denseBuilder) {
+  return denseBuilder(llvm::makeArrayRef(data.begin(), data.end()));
+}
+
+// When the protobuf repeated field has a type larger than T,
+// copy the data into correctly typed SmallVector because
+// DenseElementsAttr needs argument type of the correct bitwidth.
+template <typename T, typename U>
+std::enable_if_t<!std::is_same_v<T, U>, mlir::DenseElementsAttr>
+createDenseElmAttrFromProtoData(const google::protobuf::RepeatedField<U> &data,
+    DenseElementsAttrBuilder<T> denseBuilder) {
+  llvm::SmallVector<T> copy(data.begin(), data.end());
+  return denseBuilder(llvm::makeArrayRef(copy));
+}
+
 // Returns DenseElementsAttr with tp's data.
 template <typename T>
 mlir::DenseElementsAttr createDenseElmAttr(const std::string &externalDataDir,
@@ -148,13 +168,7 @@ mlir::DenseElementsAttr createDenseElmAttr(const std::string &externalDataDir,
   } else {
     // Not raw, no need to take care of endianness.
     const auto &data = TransformValueToONNXData<T>::data(tp);
-    // Access data directly via ArrayRef if same size as T,
-    // or copy into correctly typed SmallVector otherwise
-    // because DenseElementsAttr needs argument type of the correct bitwidth.
-    typedef typename std::conditional<sizeof(T) == sizeof(data[0]),
-        llvm::ArrayRef<T>, llvm::SmallVector<T>>::type ArrayRefOrSmallVector;
-    ArrayRefOrSmallVector array(data.begin(), data.end());
-    return denseBuilder(llvm::makeArrayRef(array));
+    return createDenseElmAttrFromProtoData(data, denseBuilder);
   }
 }
 

